pkg/dockerregistry/server/repositorymiddleware.go

```diff
@@ -372,29 +372,45 @@ func (r *repository) getImage(dgst digest.Digest) (*imageapi.Image, error) {
 	return image, nil
 }
 
-// getImageOfImageStream retrieves the Image with digest `dgst` for the ImageStream associated with r. This
-// ensures the image belongs to the image stream. It uses two queries to master API:
+// getStoredImageOfImageStream retrieves the Image with digest `dgst` and
+// ensures that the image belongs to the ImageStream associated with r. It
+// uses two queries to master API:
 //  1st to get a corresponding image stream
 //  2nd to get the image
 // This allows us to cache the image stream for later use.
-func (r *repository) getImageOfImageStream(dgst digest.Digest) (*imageapi.Image, *imageapi.ImageStream, error) {
+func (r *repository) getStoredImageOfImageStream(dgst digest.Digest) (*imageapi.Image, *imageapi.TagEvent, *imageapi.ImageStream, error) {
 	stream, err := r.imageStreamGetter.get()
 	if err != nil {
 		context.GetLogger(r.ctx).Errorf("failed to get ImageStream: %v", err)
-		return nil, nil, wrapKStatusErrorOnGetImage(r.name, dgst, err)
+		return nil, nil, nil, wrapKStatusErrorOnGetImage(r.name, dgst, err)
 	}
 
-	_, err = imageapi.ResolveImageID(stream, dgst.String())
+	tagEvent, err := imageapi.ResolveImageID(stream, dgst.String())
 	if err != nil {
 		context.GetLogger(r.ctx).Errorf("failed to resolve image %s in ImageStream %s/%s: %v", dgst.String(), r.namespace, r.name, err)
-		return nil, nil, wrapKStatusErrorOnGetImage(r.name, dgst, err)
+		return nil, nil, nil, wrapKStatusErrorOnGetImage(r.name, dgst, err)
 	}
 
 	image, err := r.getImage(dgst)
 	if err != nil {
-		return nil, nil, wrapKStatusErrorOnGetImage(r.name, dgst, err)
+		return nil, nil, nil, wrapKStatusErrorOnGetImage(r.name, dgst, err)
+	}
+
+	return image, tagEvent, stream, nil
+}
+
+// getImageOfImageStream retrieves the Image with digest `dgst` for
+// the ImageStream associated with r.
+func (r *repository) getImageOfImageStream(dgst digest.Digest) (*imageapi.Image, *imageapi.ImageStream, error) {
```

Comment:
Add a fat warning that the returned image doesn't contain original `dockerImageReference` and thus it's not suitable for sending back to master API.

Also describe what is the difference between this and the method above.